src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs

```diff
@@ -174,49 +176,92 @@ private static Task<HttpResponseMessage> InnerSendAsync(HttpRequestMessage reque
+                    case AuthenticationType.Digest:
+                        var digestResponse = new DigestResponse(challenge.ChallengeData);
+                        if (await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isProxyAuth).ConfigureAwait(false))
                         {
-                            digestResponse = new DigestResponse(challengeData);
-                            if (IsServerNonceStale(digestResponse) &&
-                                await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isProxyAuth).ConfigureAwait(false))
+                            response.Dispose();
+                            response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+
+                            // Retry in case of nonce timeout in server.
+                            if (TryGetRepeatedChallenge(response, challenge.SchemeName, isProxyAuth, out string challengeData))
                             {
-                                response.Dispose();
-                                response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+                                digestResponse = new DigestResponse(challengeData);
+                                if (IsServerNonceStale(digestResponse) &&
+                                    await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isProxyAuth).ConfigureAwait(false))
+                                {
+                                    response.Dispose();
+                                    response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
+                                }
                             }
                         }
-                    }
-                }
-                else if (challenge.AuthenticationType == AuthenticationType.Basic)
-                {
-                    if (!preAuthenticate)
-                    {
-                        SetBasicAuthToken(request, challenge.Credential, isProxyAuth);
+                        break;
+
+                    case AuthenticationType.Basic:
+                        if (performedBasicPreauth)
+                        {
+                            break;
+                        }
 
                         response.Dispose();
+                        SetBasicAuthToken(request, challenge.Credential, isProxyAuth);
                         response = await InnerSendAsync(request, isProxyAuth, doRequestAuth, pool, cancellationToken).ConfigureAwait(false);
-                    }
+
+                        if (preAuthenticate)
+                        {
+                            switch (response.StatusCode)
+                            {
+                                case HttpStatusCode.ProxyAuthenticationRequired:
```

Comment:
This seems too generous -- e.g. do we really want to interpret a 5xx as auth success?  My instinct is only 2xx/3xx should be interpreted that way.  Not sure what other handlers do here though.

Reply:
> I'll add a test for a bunch of different status codes and update it if need be.

As we've found previously any time we look at something like this, there are differences between the various handlers.

I tried all of the status codes from 100 to 599....

NetfxHandler treated all of them as success for auth with the exception of
- Unauthorized (401), as expected, and
- Continue (100)

WinHttpHandler treated all of them as success for auth, with the exception of:
- All codes in the range [100, 199]
- Moved (301)
- Redirect (302)
- RedirectMethod (303)
- TemporaryRedirect (307)
- PermanentRedirect (308)
- Unauthorized (401)
- ProxyAuthenticationRequired (407)

CurlHandler treated all of them as success for auth, with the exception of:
- Unauthorized (401)

Given this, I think more closely matching NetfxHandler and CurlHandler, i.e. what's in this PR, makes sense.  If nothing else, it's the easiest to explain.  But, I don't have a strong opinion.

Reply:
OK

Reply:
>I think more closely matching NetfxHandler

Yes, we should match .NET Framework as much as possible. That HTTP stack has been in use for 15+ years and its behaviors are what most devs are expecting when moving to .NET Core.